Add TypeScript usage example (#187)
HipsterBrown/xs-dev: docs/src/content/docs/guide/usage-with-typescript.mdx

---
title: Usage with TypeScript
description: Initialize a new project with TypeScript and start to interact with some hardware!
---

import { Aside } from '@astrojs/starlight/components'

Using [TypeScript](https://www.typescriptlang.org/) is a industry standard in 2024. It provides a lot of benefits over JavaScript, such as static type checking, better code completion, and improved code readability. This guide will show you how to use TypeScript with the Moddable SDK.

# Prerequisites

There are various ways to start a new project with TypeScript. But first let's detail what is needed for the Moddable SDK.

> A moddable project should ideally lives within a single file. This is because the Moddable SDK does not support the `import` and `export` keywords. This means that you will need to use a bundler to compile your TypeScript code into a single file. Or you could relies on `mcpack` and the dependencies management system from Moddable but this get a bit complicated.

Author: Yes, I get it better now. What I wanted to write was that if you don't use a bundler, as there is no relative file-system (as far as i understand what's going on), you need to declare things in `manifest.json` to allows mcpack to found those deps, which is a bit strange coming from a Web background. I'll try to improve this paragraph taking in consideration your notes. And submit changes. (thanks for reviewing !)

Two side notes
- I should read `mcpack` doc in detail, I've skipped the doc about it too much I guess.
- This is really sad that https://github.com/phoddie/node-cron and other things like these aren't indexed somewhere, this is pure gold. Maybe we could push them to my newly claimed `@embedded-js` org on npm ? What do you think ?

Using a bundler is very popular in the JavaScript community especially in the web development where you want to reduce the number of network request and optimize the JS bundle server. There are many tools available to help you with this task. Tools like Webpack, Rollup, and EsBuild are popular choices for bundling JavaScript code.

## Let's get started

For the purpose of this guide, as bundler we will use [Bun](https://bun.sh/), which is a JavaScript Environment, but also a bundler/transpiler. It is a very simple tool that allows you to write your code in TypeScript and bundle it into a single file. If you use [mise-en-place](https://mise.jdx.dev/getting-started.html), you can install it with the following command:

```bash
mise install bun
```

Otherwise follow the instructions on the [Bun website](https://bun.sh/).

Now we will use a starter template that has setup everything for us upfront.

```bash

> git clone git@github.com:ScreamZ/xs-dev-Typescript-template.git

Author: Sounds an exciting tool ! Wasn't aware of its existence ! Do you think suggesting it through `bunx` is relevant here? Maybe `npx` would be more familial to dummies ?

```

The documentation of this template is self-explanatory, but we will go through it in this guide for the details.

## Specializing the starter template for your project

The first thing you need to do is to update the `package.json` file with your project details. You can do this by running the following command:

1. Update `name` property with your project name.
2. Update `author` property with your name.
3. Remove existing git repository and add your own.

```bash
rm -rf .git
git init
```

4. Follow the remaining instruction on the github [https://github.com/ScreamZ/xs-dev-Typescript-template](https://github.com/ScreamZ/xs-dev-Typescript-template)

## How does it work?

The template is a simple project that uses Bun to bundle the TypeScript code into a single file. The TypeScript code is located in the `src` folder, and the bundled code is located in the `dist` folder.

> The starter exposes a `pnpm dev` command that watch for changes in the `src` folder and automatically recompile the code. This is very useful when you are developing your project. You should keep this command running in a terminal window while you are working on your project.

Author: I don't know… What's really nice with `pnpm` is that they bring this shared store that reduce size on disk. This is also a widely adopted and mature tool.

But there are still many users on yarn 1, (not many on yarn berry, as its basically a fail on doing what pnpm achieved)

Some are trying bun package manager, it's nice but still bun is relatively new.

IMO, I would apply on adding pnpm. this is not a bad choice and a drop-in replacement for npm.

When you need to try the code, simply run `xs-dev run` in the terminal. This will upload the bundled code from `dist/main.js` to the device and run it. The rest remain the same as the standard xs-dev workflow.

Take note that whenever you import some code from the Moddable SDK, you will need to add the following line in the `tsconfig.json` file:

```json
"types": [
     // ... Previous lines
      "./node_modules/@moddable/typings/<path-to-the-file>"
    ]
```

Those typings are provided by `@moddable/typings` package, and while at the moment this get a bit cumbersome but it's up to you to find the correct definition file for the module you are using.

As an example Timer and WiFi lives in

````
      "./node_modules/@moddable/typings/wifi",
      "./node_modules/@moddable/typings/timer",
      ```
````

The team is actively working on improving the development workflow.

<Aside>

>   As modules from ModdableSDK are not available inside node_modules and injected

Reviewer: nit: missing space

```suggestion
  As modules from the Moddable SDK are not available inside node_modules and injected
```

---
title: Usage with TypeScript
description: Initialize a new project with TypeScript and start to interact with some hardware!
---

import { Aside } from '@astrojs/starlight/components'

> Using [TypeScript](https://www.typescriptlang.org/) is a industry standard in 2024. It provides a lot of benefits over JavaScript, such as static type checking, better code completion, and improved code readability. This guide will show you how to use TypeScript with the Moddable SDK.

Reviewer: ```suggestion
Using [TypeScript](https://www.typescriptlang.org/) is an industry standard in 2024. It provides a lot of benefits over JavaScript, such as static type checking, better code completion, and improved code readability. This guide will show you how to use TypeScript with the Moddable SDK.
```

## Let's get started

This starter project aim to build a single JS file that will be passed to XS. This helps developers coming from the web to have a more familiar environment and also to have a better development experience.

For this purpose, we will use a bundler tool which is very popular in the JavaScript community, especially in the web development where you want to reduce the number of network request and optimize the JS bundle from the server. There are many tools available to help you with this task. Tools like Webpack, Rollup, and esbuild are popular choices for bundling JavaScript code.
For this guide, as the bundler we will use [Bun](https://bun.sh/), which is a JavaScript Environment, but also a bundler/transpiler. It is a very simple tool that allows you to write your code in TypeScript and bundle it into a single file. If you use [mise-en-place](https://mise.jdx.dev/getting-started.html), you can install it with the following command:

Along with Bun, we will add Typescript as a dev dependency because bundler for performances reasons do not lint the code. Therefore we will use `tsc` for that.

```bash
mise install bun
```

> Otherwise follow the instructions on the [Bun website](https://bun.sh/).

Reviewer: To help with readability, we should display the mise install command after the sentence "you can install it with the following command:"

~~~suggestion
For this guide, as the bundler we will use [Bun](https://bun.sh/), which is a JavaScript Environment, but also a bundler/transpiler. It is a very simple tool that allows you to write your code in TypeScript and bundle it into a single file. If you use [mise-en-place](https://mise.jdx.dev/getting-started.html), you can install it with the following command:

```bash
mise install bun
```

Otherwise follow the instructions on the [Bun website](https://bun.sh/).

Along with Bun, we will add Typescript as a dev dependency because bundler for performances reasons do not lint the code. Therefore we will use `tsc` for that.
~~~

I don't see any instructions for adding TypeScript as a dev dependency. If you're referring to the starter template including TypeScript as a dev dependency, we should be specific about that.

Now we will use a [starter template](https://github.com/ScreamZ/xs-dev-Typescript-template) that has setup everything for us upfront.

```bash
 bunx tiged ScreamZ/xs-dev-TypeScript-template ts-xs-project
```

<Aside>
  The `bunx` command is a wrapper around `git clone` that will clone the
  repository and remove the `.git` folder. This is useful when you want to start

>   a new project from a template.

Reviewer: `tiged` is the CLI that wraps around `git clone`, while `bunx` allows for running package commands without globally installing said package (similar to `npx`, `pnpx`, and `yarn dlx`.

</Aside>

## Specializing the starter template for your project

> The first thing you need to do is to update the `package.json` file with your project details. You can do this by running the following command:

Reviewer: Is there a command to run? Or just a set of instructions?

Author: this is a set of instructions now, forgot to update after the removal of things in favor of tiged

1. Update `name` property with your project name.
2. Update `author` property with your name.
3. Follow the remaining instruction on the GitHub [https://github.com/ScreamZ/xs-dev-Typescript-template](https://github.com/ScreamZ/xs-dev-Typescript-template)

## How does it work?

The template is a simple project that uses Bun to bundle the TypeScript code into a single file. The TypeScript code is located in the `src` folder, and the bundled code is located in the `dist` folder.

The starter exposes a `pnpm dev` command that watch for changes in the `src` folder and automatically recompile the code. This is very useful when you are developing your project. You should keep this command running in a terminal window while you are working on your project.

> When you need to try the code, simply run `xs-dev run` in the terminal. This will upload the bundled code from `dist/main.js` to the device and run it. The rest remain the same as the standard xs-dev workflow.

Reviewer: nit: grammar

```suggestion
When you need to try the code, simply run `xs-dev run` in the terminal. This will upload the bundled code from `dist/main.js` to the device and run it. The rest remains the same as the standard xs-dev workflow.
```

Take note that whenever you import some code from the Moddable SDK, you will need to add the following line in the `tsconfig.json` file:

```json
"types": [
     // ... Previous lines
      "./node_modules/@moddable/typings/<path-to-the-file>"
    ]
```

> Those typings are provided by [`@moddable/typings` package](https://github.com/Moddable-OpenSource/moddable/tree/4142d4d353960ea7ede15d56864d841de692a68d/typings), and while at the moment this get a bit cumbersome but it's up to you to find the correct definition file for the module you are using.

Reviewer: ```suggestion
Those typings are provided by [`@moddable/typings` package](https://github.com/Moddable-OpenSource/moddable/tree/4142d4d353960ea7ede15d56864d841de692a68d/typings), and while at the moment this is a bit cumbersome, it's up to you to find the correct definition file for the module you are using.
```

> As an example Timer and WiFi lives in `./node_modules/@moddable/typings/wifi` and `./node_modules/@moddable/typings/timer`

Reviewer: ```suggestion
As an example, the Timer and WiFi modules live in `./node_modules/@moddable/typings/wifi` and `./node_modules/@moddable/typings/timer`
```
